### PR TITLE
Clean up STJ references Fixes #39902

### LINKED ISSUE #39902: Clean up our System.Text.Json references
In digging through https://github.com/dotnet/sdk/pull/39901/files, we realized that our logic for UseSystemTextJson is a bit off.

We now use system.text.json even on framework builds. I believe @ladipro has a change to make this even more common. On framework, we need to ensure we have an 8.0.0 reference.

For core, we shouldn't have a reference at all (but we add it in a couple of places).

Changes needed:
1. Ensure all projects that run in framework and use System.Text.Json have an explicit 8.0.0 reference
2. Put that version in the versions.props file so it's easier to update in main
  a. It appears to be in main already but unused: https://github.com/dotnet/sdk/blob/main/eng/Versions.props#L38
3. remove the reference in .net core builds where we're accidentally adding it (search for UseSystemTextJson and review the logic there)

### PR DESCRIPTION
**Pull request by Forgind**
Fixes #39902

First commit removes references to S.T.J that shouldn't be necessary because we have our own "WorkloadManifestReader.SystemTextJson" class.

The second commit forces all our package versions to 8.0.0.

I was a bit confused with this logic:
`<UseSystemTextJson Condition="'$(TargetFramework)'!='netstandard2.0' And '$(TargetFramework)'!='net472'">True</UseSystemTextJson>`

This seems to say that we should use System.Text.Json on core only

### CODE REVIEW

File: src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/Microsoft.DotNet.MSBuildSdkResolver.csproj [hunk 1]
```diff
@@ -91,7 +91,7 @@
   </ItemGroup>
 
   <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
-    <PackageReference Include="System.Text.Json"  VersionOverride="8.0.0" />
+    <PackageReference Include="System.Text.Json"  VersionOverride="$(SystemTextJsonPackageVersion)" Condition="'$(TargetFramework)' == 'net472'"/>
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/Microsoft.DotNet.MSBuildSdkResolver.csproj:94]: I'm just confused why things aren't working the way I'd expect, so I was adding some duplication in case I had missed something, and it didn't work as I'd thought.

File: eng/Versions.props [hunk 2]
```diff
@@ -78,7 +78,7 @@
     <SystemSecurityCryptographyXmlPackageVersion>8.0.0</SystemSecurityCryptographyXmlPackageVersion>
     <SystemSecurityPermissionsPackageVersion>8.0.0</SystemSecurityPermissionsPackageVersion>
     <SystemServiceProcessServiceControllerVersion>8.0.0</SystemServiceProcessServiceControllerVersion>
-    <SystemTextJsonPackageVersion>8.0.2</SystemTextJsonPackageVersion>
+    <SystemTextJsonPackageVersion>8.0.0</SystemTextJsonPackageVersion>
```
Comment (marcpopMSFT) [inline eng/Versions.props:81]: This property is used in directory.packages.props so what you're doing here is actually overriding the version used by transitive pinning (so overriding the version for all implicit .net core references). You'll need a separate property to use for pinning to 8.0.0 and let the d.p.props version float to current.  Runtime does this too if you see here: https://github.com/dotnet/runtime/blob/main/eng/Versions.props#L137

FYI, I talked to Eric and I'll be prepping a change for 8.0.1xx and 8.0.3xx for how we handle STJ. We'll want that change in 8.0.4xx as well. It depends on a change in runtime so you won't be able to get this PR working in 4xx as it turns out. You should be able to retarget to main, make the above change, apply my changes, and get it sorted out. I'll talk to you offline as it's complicated as there are actually 3 different classes of STJ references, not 2 like I thought.

Comment (marcpopMSFT) [reply eng/Versions.props:81]: Following up on this, I believe there are three categories of project that we talked about:
1. The four task projects should have STJ references with no version override. Razor and static web assets should have a net472 check but build and publish should not. I recommend looking at the 8.0.3xx internal AzDo branches to confirm:
https://dev.azure.com/dnceng/internal/_git/dotnet-sdk?path=/src/WebSdk/Publish/Tasks/Microsoft.NET.Sdk.Publish.Tasks.csproj&version=GBinternal/release/8.0.3xx&_a=contents
2. The components we have that build for framework and run in VS should have a reference to STJ for netfx targeting and should have a VersionOverride to version 8.0.0. That's because in these cases, VS is providing STJ and we need to work with whichever version they ship.
3. No other .net core building projects should have a reference to STJ.

CC @ericstj to confirm. We discussed this in a 1:1 a few weeks ago but didn't write it down so writing it down here now.

Comment (ericstj) [reply eng/Versions.props:81]: This sounds correct.  Minimize the places where you ship a copy if you can reference the version provided by your host process.

File: src/BuiltInTools/dotnet-watch/dotnet-watch.csproj [hunk 3]
```diff
@@ -33,11 +33,6 @@
     <PackageReference Include="Microsoft.CodeAnalysis.Workspaces.MSBuild" />
     <PackageReference Include="Microsoft.NET.StringTools" ExcludeAssets="runtime" PrivateAssets="All" />
 
-    <!--
-      Explicit System.Text.Json package reference is required for source-build to pick up the live package.
-      Avoids picking up an old version via transitive dependency from Microsoft.Build or Microsoft.CodeAnalysis.Workspaces.MSBuild.
-    -->
-    <PackageReference Include="System.Text.Json" />
```
Comment (marcpopMSFT) [inline src/BuiltInTools/dotnet-watch/dotnet-watch.csproj:40]: Probably want a review from @tmat but with CPM now enabled, this may not be required anymore like it was before.

Comment (Forgind) [reply src/BuiltInTools/dotnet-watch/dotnet-watch.csproj:40]: Having an explicit package reference is no longer required, or removing it is no longer required?

File: src/Tests/ArgumentsReflector/ArgumentsReflector.csproj [hunk 4]
```diff
@@ -23,6 +23,7 @@
     <Using Remove="Microsoft.NET.TestFramework.Utilities" />
     <Using Remove="Xunit" />
     <Using Remove="Xunit.Abstractions" />
+    <PackageReference Include="System.Text.Json" VersionOverride="$(SystemTextJsonPackageVersion)" />
```
Comment (marcpopMSFT) [inline src/Tests/ArgumentsReflector/ArgumentsReflector.csproj:26]: this is a test so not sure it'll need the version override

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/Microsoft.NET.Sdk.WorkloadManifestReader.csproj [hunk 5]
```diff
@@ -25,7 +25,9 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="System.Text.Json" Condition="'$(UseSystemTextJson)'=='True'" />
+    <PackageReference Include="System.Text.Json"
```
Comment (marcpopMSFT) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/Microsoft.NET.Sdk.WorkloadManifestReader.csproj:28]: This one will require more investigation. The conditional above that sets UseSystemTextJson is configured for core only which we don't need a reference for so I think we might be able to just remove it from here but you'd have to confirm the framework build doesn't use STJ (as I know things changed so this property may not be accurately named anymore)

File: src/Cli/Microsoft.DotNet.Cli.Utils/Microsoft.DotNet.Cli.Utils.csproj [hunk 6]
```diff
@@ -47,6 +47,7 @@
     <PackageReference Include="Microsoft.Build" ExcludeAssets="runtime" PrivateAssets="all" />
     <PackageReference Include="Microsoft.Build.Utilities.Core" ExcludeAssets="runtime" PrivateAssets="all" />
     <PackageReference Include="System.CommandLine" />
+    <PackageReference Include="System.Text.Json" VersionOverride="$(SystemTextJsonPackageVersion)" Condition="'$(TargetFramework)' == 'net472'"/>
```
[the remark this answers is not in the document]
Comment (marcpopMSFT) [reply src/Cli/Microsoft.DotNet.Cli.Utils/Microsoft.DotNet.Cli.Utils.csproj:50]: I'm starting to think the PR to create the specific package versions was never ported to main. I'll need to investigate.

File: src/BuiltInTools/dotnet-watch/dotnet-watch.csproj (hunk 3 above)
Comment (marcpopMSFT) [reply src/BuiltInTools/dotnet-watch/dotnet-watch.csproj:40]: Having the explicit reference is no longer required, I think. If I understand the original issue per the comment, a transitive version was picked up that didn't work with source build and adding the explicit reference up-versioned to one that was supported. CPM should do the version change as well.

File: src/Cli/Microsoft.DotNet.Cli.Utils/Microsoft.DotNet.Cli.Utils.csproj (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcpopMSFT) [reply src/Cli/Microsoft.DotNet.Cli.Utils/Microsoft.DotNet.Cli.Utils.csproj:50]: My PR is now in which should enable you to update the minimum version name #41890

File: src/Tests/ArgumentsReflector/ArgumentsReflector.csproj [hunk 7]
```diff
@@ -23,6 +23,7 @@
     <Using Remove="Microsoft.NET.TestFramework.Utilities" />
     <Using Remove="Xunit" />
     <Using Remove="Xunit.Abstractions" />
+    <PackageReference Include="System.Text.Json" />
```
Comment (marcpopMSFT) [inline src/Tests/ArgumentsReflector/ArgumentsReflector.csproj:26]: Do you know why this was needed?

File: src/Cli/Microsoft.DotNet.Cli.Utils/Microsoft.DotNet.Cli.Utils.csproj [hunk 8]
```diff
@@ -47,6 +47,7 @@
     <PackageReference Include="Microsoft.Build" ExcludeAssets="runtime" PrivateAssets="all" />
     <PackageReference Include="Microsoft.Build.Utilities.Core" ExcludeAssets="runtime" PrivateAssets="all" />
     <PackageReference Include="System.CommandLine" />
+    <PackageReference Include="System.Text.Json" VersionOverride="$(SystemTextJsonBasePackageVersion)" Condition="'$(TargetFramework)' == 'net472'"/>
```
Comment (marcpopMSFT) [inline src/Cli/Microsoft.DotNet.Cli.Utils/Microsoft.DotNet.Cli.Utils.csproj:50]: why was this needed?

**Review by marcpopMSFT — APPROVED** [review]

File: src/Microsoft.DotNet.TemplateLocator/Microsoft.DotNet.TemplateLocator.csproj [hunk 9]
```diff
@@ -25,7 +25,7 @@
       <BuildOutputInPackage Include="@(ReferenceCopyLocalPaths-&gt;WithMetadataValue('ReferenceSourceTarget', 'ProjectReference'))" />
     </ItemGroup>
   </Target>
-
+
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.TemplateLocator/Microsoft.DotNet.TemplateLocator.csproj:28]:
```suggestion

```

File: eng/Versions.props [hunk 10]
```diff
@@ -157,6 +157,7 @@
     <SystemSecurityPermissionsPackageVersion>9.0.0-preview.7.24366.18</SystemSecurityPermissionsPackageVersion>
     <SystemTextEncodingCodePagesPackageVersion>9.0.0-preview.7.24366.18</SystemTextEncodingCodePagesPackageVersion>
     <SystemTextJsonPackageVersion>9.0.0-preview.7.24366.18</SystemTextJsonPackageVersion>
+    <SystemTextJsonBasePackageVersion>8.0.0</SystemTextJsonBasePackageVersion>
```
Comment (ViktorHofer) [inline eng/Versions.props:160]: What does "Base" stand for here? You might want to add a comment.

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/Microsoft.NET.Sdk.WorkloadManifestReader.csproj [hunk 11]
```diff
@@ -28,7 +28,6 @@
   <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
     <!-- Netfx version builds against the lowest version of System.Text.Json that's guaranteed to be shipped with MSBuild in VS -->
     <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" VersionOverride="8.0.0" />
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/Microsoft.NET.Sdk.WorkloadManifestReader.csproj:30]: I was confused in looking at the 8.0.4xx-targeted version, as this wasn't needed then...but that's because it was using Newtonsoft.Json instead. Apparently it was fully switched to STJ, so now it became a problem. Thanks for catching that 🙂

File: eng/Versions.props [hunk 12]
```diff
@@ -157,6 +157,9 @@
     <SystemSecurityPermissionsPackageVersion>9.0.0-preview.7.24366.18</SystemSecurityPermissionsPackageVersion>
     <SystemTextEncodingCodePagesPackageVersion>9.0.0-preview.7.24366.18</SystemTextEncodingCodePagesPackageVersion>
     <SystemTextJsonPackageVersion>9.0.0-preview.7.24366.18</SystemTextJsonPackageVersion>
+    <!-- This is a minimum version for various projects to work. It's used for netfx-targeted components that run in Visual Studio
+         because in those cases, Visual Studio is providing System.Text.Json, and we should work with whichever version it ships. -->
+    <SystemTextJsonBasePackageVersion>8.0.0</SystemTextJsonBasePackageVersion>
```
Comment (Forgind) [inline eng/Versions.props:162]: Need to update to 8.0.4
```suggestion
    <SystemTextJsonBasePackageVersion>8.0.4</SystemTextJsonBasePackageVersion>
```

File: eng/Versions.props [hunk 13]
```diff
@@ -157,6 +157,9 @@
     <SystemSecurityPermissionsPackageVersion>9.0.0-preview.7.24366.18</SystemSecurityPermissionsPackageVersion>
     <SystemTextEncodingCodePagesPackageVersion>9.0.0-preview.7.24366.18</SystemTextEncodingCodePagesPackageVersion>
     <SystemTextJsonPackageVersion>9.0.0-preview.7.24366.18</SystemTextJsonPackageVersion>
+    <!-- This is a minimum version for various projects to work. It's used for netfx-targeted components that run in Visual Studio
+         because in those cases, Visual Studio is providing System.Text.Json, and we should work with whichever version it ships. -->
+    <SystemTextJsonBasePackageVersion>8.0.4</SystemTextJsonBasePackageVersion>
```
Comment (rainersigwald) [inline eng/Versions.props:162]: Bumping the netfx reference to 8.0.4 will fail to satisfy this constraint for VS 17.11. That's bad, right?

**Comment by Forgind** [comment]
The ubuntu failure appears to be https://github.com/dotnet/source-build/issues/4503

**Review by marcpopMSFT — APPROVED** [review]
